### PR TITLE
Release unreleased revisions

### LINKED ISSUE #1069: Release unassigned revison

### PR DESCRIPTION
**Pull request by bartaz**
Fixes #1069 

Adds ability to release unreleased revisions (from Revision history table) to edge

### QA

- ./run or http://snapcraft.io-pr-1085.run.demo.haus/
- go to Release page of snap you own
- Click on arrow (Promote) button next to revision from revisions table
- it should be displayed in releases table as pending release
- try to cancel it, or promote to other channel

screen shot 2018-09-17 at 15 44 11

### CODE REVIEW

**Comment by therealjuan** [comment]
A few comments:

- Can we include a `Promote to {TRACK_NAME}/edge` tooltip on unreleased revisions?
- Can we include a `Promote to {TRACK_NAME}/{CHANNEL_NAME}` tooltip when promoting up to stable (not included)?
- Have a tooltip on the `X` that says 'Revert promoting this channel`?
- Is there any way to increase the threshold for revisions listed? In a snap like `nextcloud` it might be tricky as some of the revisions are not displayed currently so it is hard to relate from the channel map to the table

File: static/js/publisher/release.js [hunk 1]
```diff
@@ -97,16 +96,14 @@ const initReleases = (id, snapName, releasesData, channelMapsList, options) => {
...
-      <ReleasesController
-        snapName={snapName}
-        releasedChannels={releasedChannels}
-        tracks={tracks}
-        archs={archs}
-        options={options}
-      />
-      <RevisionsList revisions={releasesData.revisions} />
-    </Fragment>,
+    <ReleasesController
```
Comment (Lukewh) [inline static/js/publisher/release.js]: nice

File: static/js/publisher/release/releasesController.js [hunk 2]
```diff
@@ -3,13 +3,17 @@ import PropTypes from 'prop-types';
...
+import RevisionsList from './revisionsList';
 import Notification from './notification';
 
 export default class ReleasesController extends Component {
   constructor(props) {
     super(props);
 
     this.state = {
+      // default to latest track
+      currentTrack: 'latest',
```
Comment (Lukewh) [inline static/js/publisher/release/releasesController.js]: we need to figure out a way to set a 'default' before it's added to the rest of the snap stack - this won't have the desired effect for node. Maybe an exceptions mapping flask-side that is exposed to all templates, so the default 'default' is 'latest' but we can override.

File: static/js/publisher/release/releasesController.js [hunk 3]
```diff
@@ -3,13 +3,17 @@ import PropTypes from 'prop-types';
...
+
```
Comment (Lukewh) [inline static/js/publisher/release/releasesController.js]: :totoemoji: - space

File: static/js/publisher/release/releasesController.js [hunk 4]
```diff
@@ -27,10 +31,28 @@ export default class ReleasesController extends Component {
...
     this.setState((state) => {
       const { pendingReleases } = state;
 
+      // cancel any other pending release for the same channel in same architectures
+      revision.architectures.forEach((arch) => {
+        Object.keys(pendingReleases).forEach((revisionId) => {
+          const pendingRelease = pendingReleases[revisionId];
+
+          if (
+            pendingRelease.channels.indexOf(channel) !== -1 &&
```
Comment (Lukewh) [inline static/js/publisher/release/releasesController.js]: `.includes` maybe instead?

File: static/js/publisher/release/releasesController.js [hunk 5]
```diff
@@ -205,17 +227,26 @@ export default class ReleasesController extends Component {
...
+          currentTrack={this.state.currentTrack}
           tracks={tracks}
           archs={archs}
-          options={options}
-          releaseRevisions={this.releaseRevisions.bind(this)}
-          fetchStatus={this.state}
-
+          isLoading={this.state.isLoading}
           pendingReleases={this.state.pendingReleases}
+
```
Comment (Lukewh) [inline static/js/publisher/release/releasesController.js]: :totoemoji: - space

File: static/js/publisher/release/revisionsList.js [hunk 6]
```diff
@@ -21,11 +35,42 @@ export default class RevisionsList extends Component {
...
+  releaseClick(revision) {
+    this.props.promoteRevision(revision, `${this.props.currentTrack}/edge`);
+  }
+
+  undoClick(revision) {
+    this.props.undoRelease(revision, `${this.props.currentTrack}/edge`);
+  }
+
+  isAlreadyReleased(revision) {
+    return Object.keys(this.props.releasedChannels).some((channel) => {
```
Comment (Lukewh) [inline static/js/publisher/release/revisionsList.js]: `.some` - nice

**Comment by Lukewh** [comment]
screenshot_2018-09-18 releases and revision history for nextcloud

Can we add `overflow: ellipsis` or something? And show the full name in a tooltip?

**Comment by bartaz** [comment]
@Lukewh The button overlap bug is a 'known issue' - we have overflow ellipsis in place, but buttons overlap over it. The issue is that there may be one button, two or no at all and because they are positioned absolutely it's hard to limit width of the cell properly.

The visual design update for this is in works, so we will try to address it later.

File: static/js/publisher/release/revisionsList.js (hunk 6 above)
Comment (bartaz) [reply static/js/publisher/release/revisionsList.js]: Oh yeah, now I feel smart, next week I'll not remember what the hell this thing does.

File: static/js/publisher/release.js (hunk 1 above)
Comment (bartaz) [reply static/js/publisher/release.js]: Trying to stay sane without adding redux just yet...

**Pull request merged**